docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+🚫
+
+```typescript
+const tokens = []; // Type 'any[]'
+const tokensMap = new Map(); // Type 'Map<any, any>'
+```
+
+✅
+
+```typescript
+const tokens: string[] = []; // Type 'string[]'
+const tokensMap = new Map<string, Token>(); // Type 'Map<string, Token>'
+```
+
+### Type Assertions
+
+`as` assertions are inherently unsafe. They overwrite type-checked and inferred types with user-supplied types that suppress compiler errors.
+
+They should only be introduced into the code if the accurate type is unreachable through other means.
+
+##### Document safe or necessary use of type assertions
+
+When a type assertion is absolutely necessary due to constraints or is even safe due to runtime checks, we should document the reasoning behind its usage in the form of a comment.
+
+<!-- TODO: Add example -->
+
+#### Avoid `as`
+
+Type assertions make the code brittle against changes.
+
+While TypeScript and ESLint will flag some unsafe, structurally unsound, or redundant type assertions, they will generally accept user-supplied types without further type-checking.
+
+This can cause silent failures or false negatives where errors are suppressed. This is especially damaging as the codebase accumulates changes over time. Type assertions may continue to silence errors, even though the type itself, or the type's relationship to the rest of the code may have been altered so that the asserted type is no longer valid.
+
+##### Redundant or unnecessary `as` assertions are not flagged for removal
+
+Type assertions can also cause false positives, because assertions are independent expressions, untied to the type errors they were intended to fix. Even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and now should be removed.
+
+```typescript
+enum Direction {
+  Up = 'up',
+  Down = 'down',
+  Left = 'left',
+  Right = 'right',
+}
+const directions = Object.values(Direction);
+
+// Error: Element implicitly has an 'any' type because index expression is not of type 'number'.(7015)
+// Only one of the two `as` assertions necessary to fix error, but neither are flagged as redundant.
+for (const key of Object.keys(directions) as keyof directions[]) {
+  const direction = directions[key as keyof typeof directions];
+}
+```
+
+##### Type guards can be used to improve type inference and avoid type assertion
+
+```typescript
+function isSomeInterface(x: unknown): x is SomeInterface {
+  return (
+    'name' in x &&
+    typeof x.name === 'string' &&
+    'length' in x &&
+    typeof x.length === 'number'
+  );
+}
+```
+
+🚫 Type assertion
+
+```typescript
+function f(x: SomeInterface | SomeOtherInterface) {
+  console.log((x as SomeInterface).name);
```

Comment:
Nit: This might make for a more realsitic example of what not to do:
```suggestion
  if (x.name) {
    // We know that x is `SomeInterface` because `x` has a `name` but `SomeOtherInterface` does not
    console.log((x as SomeInterface).name);
  }

```